libraries/SITL/SIM_GPS.cpp

```diff
@@ -1014,6 +1017,311 @@ uint32_t GPS::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::DIFFERENTIAL_POS_PHASE_RTK);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::POSITION_METHOD_FIXED_PHASE);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::OMNISTAR_ACTIVE);
+        pos_flags_2 |= uint8_t(POS_FLAGS_2::DETERMINED_WITH_STATIC_CONSTRAINT);
+    }
+
+    const auto gps_tow = gps_time();
+    struct PACKED gsof_pos_time {
+        const uint8_t OUTPUT_RECORD_TYPE { GSOF_POS_TIME_TYPE }; 
+        const uint8_t RECORD_LEN { GSOF_POS_TIME_LEN };
+        uint32_t time_week_ms;
+        uint16_t time_week;
+        uint8_t num_sats;
+        // TODO
+        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%201
+        uint8_t pos_flags_1;
+        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%202
+        uint8_t pos_flags_2;
+        // TODO increment every time GPS is "restarted", can we use boot count?
+        uint8_t initialized_num = bootcount; 
+    } pos_time;
+    static_assert(sizeof(gsof_pos_time) - (sizeof(gsof_pos_time::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos_time::RECORD_LEN)) == GSOF_POS_TIME_LEN);
+    pos_time.pos_flags_1 = pos_flags_1;
+    pos_time.pos_flags_2 = pos_flags_2;
+    pos_time.time_week_ms = htobe32(gps_tow.ms);
+    pos_time.time_week = htobe16(gps_tow.week);
+    pos_time.num_sats = d->have_lock ? _sitl->gps_numsats[instance] : 3;
+
+    constexpr uint8_t GSOF_POS_TYPE = 0x02;
+    constexpr uint8_t GSOF_POS_LEN = 0x18;
+
+    int64_t lat_i64 { 0 };
+    const auto lat_rad = d->latitude * DEG_TO_RAD_DOUBLE;
+    memcpy(&lat_i64, &lat_rad, sizeof(lat_i64));
+
+    int64_t lng_i64 { 0 };
+    const auto lng_rad = d->longitude * DEG_TO_RAD_DOUBLE;
+    memcpy(&lng_i64, &lng_rad, sizeof(lng_i64));
+
+    int64_t alt_i64 { 0 };
+    const double alt_m = static_cast<double>(d->altitude);
+    memcpy(&alt_i64, &alt_m, sizeof(alt_i64));
+
+    // TODO fix constness 
+    // Method 1: Use designated initializer (or aggregate initialization)
+    // Not working, perhaps because PACKED?
+    // error: designated initializers cannot be used with a non-aggregate type ‘SITL::GPS::update_gsof(const SITL::GPS::gps_data*)::gsof_pos
+    // Method 2: Initialize in-class as const
+    // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96184
+    // Perhaps a bug in gcc? 
+    // error: ../../libraries/SITL/SIM_GPS.cpp:1096:24: error: use of local variable with automatic storage from containing function
+    //
+    // https://godbolt.org/z/7oeMfMzj6
+    struct PACKED gsof_pos {
+        const uint8_t OUTPUT_RECORD_TYPE { GSOF_POS_TYPE }; 
+        const uint8_t RECORD_LEN { GSOF_POS_LEN };
+        uint64_t lat;
+        uint64_t lng;
+        uint64_t alt;
+    } pos {};
+    static_assert(sizeof(pos.lat) == sizeof(lat_i64), "Unexpected double format");
+    // pos.OUTPUT_RECORD_TYPE = GSOF_POS_TYPE;
+    // pos.RECORD_LEN = GSOF_POS_LEN;
+    pos.lat = htobe64(lat_i64);
```

Comment:
```suggestion
    pack_double_into_pos_packet(&pos.lat, d->latitude * DEG_TO_RAD_DOUBLE);
```

similarly for the 2 other fields.

Reply:
Implemented, not exactly, but very close to the idea you had.